### DISCUSSION TITLE
ui.upload `stutters` with uploading large files

### DISCUSSION THREAD

**Discussion by Rjdrenth** [zauberzeug/nicegui#3220] [Q&A] [answered]
### Question

I am building an application that lets users upload multiple files in an `ui.upload(..)` component. These files can be anywhere from 1mb to 100mb with a total upload size from 1mb to possibly 1gb. 
When this happens some form of 'upload stuttering' appears, meaning that the first part of uploading goes pretty fast (first 20 to 30%), but then hangs for a moment and then repeatedly progresses a few %, then pauses a bit etc until the file(s) are finally uploaded.

I tried attaching a short video of the phenomenon (possible sound warning, apparently it recorded some noise): 
https://github.com/zauberzeug/nicegui/assets/6966290/fcba602e-2f8d-4977-891a-52201e8f4107

This happens regardless of whether I run the application locally or at my target deployment platform (Azure Web App) or whether I have an on_upload callback (which seems logical, but I tested it anyway), so it does not seem like a network speed issue. 

I would appreciate any pointers to how I could improve the upload speed.

**Comment by Rjdrenth** [comment, marked as answer]
Okay, so in the depths of the Starlette library the [MultiPartParser](https://github.com/encode/starlette/blob/5a1bec33f8d6a669a3670f51034de83292d19408/starlette/formparsers.py#L120) is defined, which defines part of the underlying mechanism of `ui.upload`. This class has a static field `max_file_size = 1024 * 1024`, which basically tells the `SpooledTemporaryFile` (which is how uploaded files get stored) that when uploaded files are larger than 1Mb, they should be written to a temporary file on the disk, rather than being kept in RAM memory. 

In my situation this results in the creation of 1 or multiple temp files which repeatedly get written and extended when new chunks of the files get uploaded.

The solution for me to resolve the issue of stuttering uploads, is to include the following piece of code somewhere near when my application starts.
```py
from starlette.formparsers import MultiPartParser

MultiPartParser.max_file_size = 1024 * 1024 * 512
```
This increases the max_size before a file gets written to disk to 512Mb and solves my issue. 

Perhaps it might be an idea to include this in the documentation of `ui.upload` or perhaps even create a NiceGUI-wide setting (for visibility) which in turn overwrites this value in Starlette?

**Reply by falkoschindler** [reply]
Interesting! Would you like to create a pull request to add this insight to the documentation, @Rjdrenth? 🙂

**Reply by Rjdrenth** [reply]
Sure thing! I will do it some time this week.

**Reply by falkoschindler** [reply]
See PR #3235. ✔️
